src/category.rs

```diff
@@ -44,6 +45,53 @@ impl Category {
             category: category,
         }
     }
+
+    pub fn update_crate(conn: &GenericConnection,
+                        krate: &Crate,
+                        categories: &[String]) -> CargoResult<()> {
+        let old_categories = try!(krate.categories(conn));
+        let old_categories = old_categories.iter().map(|cat| {
+            (&cat.category[..], cat)
+        }).collect::<HashMap<_, _>>();
+        // If a new category specified is not in the database, filter
+        // it out and don't add it.
+        let new_categories = try!(categories.iter().filter_map(|c| {
+            let cat = Category::find_by_category(conn, &c);
+            if let Ok(Some(db_cat)) = cat {
+                Some(Ok((&c[..], db_cat)))
+            } else {
+                None
+            }
+        }).collect::<CargoResult<HashMap<_, _>>>());
+
+        let to_rm = old_categories.iter().filter(|&(cat, _)| {
+            !new_categories.contains_key(cat)
+        }).map(|(_, v)| v.id).collect::<Vec<_>>();
+        let to_add = new_categories.iter().filter(|&(cat, _)| {
+            !old_categories.contains_key(cat)
+        }).map(|(_, v)| v.id).collect::<Vec<_>>();
```

Comment:
Maybe all of this should be a `HashSet` and then use `a.difference(b)` and `b.difference(a)`?